File: Optima/LinearSolver.cpp
// Optima is a C++ library for solving linear and non-linear constrained optimization problems.
//
// Copyright © 2020-2024 Allan Leal
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more Mbar.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "LinearSolver.hpp"

// Optima includes
#include <Optima/Exception.hpp>
#include <Optima/LinearSolverFullspace.hpp>
#include <Optima/LinearSolverNullspace.hpp>
#include <Optima/LinearSolverRangespace.hpp>

namespace Optima {

struct LinearSolver::Impl
{
    LinearSolverOptions options; ///< The options for the linear solver.

    LinearSolverRangespace rangespace; ///< The linear solver based on a rangespace algorithm.
    LinearSolverNullspace nullspace;   ///< The linear solver based on a nullspace algorithm.
    LinearSolverFullspace fullspace;   ///< The linear solver based on a fullspace algorithm.

    Vector x; ///< The auxiliary solution vector x.
    Vector p; ///< The auxiliary solution vector p.
    Vector w; ///< The auxiliary solution vector w.

    Vector wbar; ///< The auxiliary solution vector w' = (w'bs, w'bu, w'bl).

    Vector ax; ///< The auxiliary solution vector ax.
    Vector aw; ///< The auxiliary solution vector aw.

    Impl()
    {}

    auto solveCanonical(CanonicalMatrix Mc, CanonicalVectorView ac, CanonicalVectorRef uc) -> void
    {
        switch(options.method)
        {
        case LinearSolverMethod::Nullspace: nullspace.solve(Mc, ac, uc); break;
        case LinearSolverMethod::Rangespace: rangespace.solve(Mc, ac, uc); break;
        default: fullspace.solve(Mc, ac, uc); break;
        }
    }

    auto decompose(CanonicalMatrix Mc) -> void
    {
        switch(options.method)
        {
        case LinearSolverMethod::Nullspace: nullspace.decompose(Mc); break;
        case LinearSolverMethod::Rangespace: rangespace.decompose(Mc); break;
        default: fullspace.decompose(Mc); break;
        }
    }

    auto solve(CanonicalMatrix Mc, MasterVectorView a, MasterVectorRef u) -> void
    {
        const auto dims = Mc.dims;
        const auto Rbs  = Mc.Rbs;
        const auto js   = Mc.js;
        const auto ju   = Mc.ju;

        const auto ns  = dims.ns;
        const auto nu  = dims.nu;
        const auto nbs = dims.nbs;
        const auto nns = dims.nns;

        ax.resize(dims.nx);
        auto as = ax.head(ns);
        auto au = ax.tail(nu);

        as.noalias() = a.x(js);
        au.noalias() = a.x(ju);

        aw.resize(dims.nw);
        auto awbs = aw.head(nbs);
        awbs = Rbs * a.w;

        const auto ap = a.p;

        solve(Mc, {as, au, ap, awbs}, u);
    }

    auto solve(CanonicalMatrix Mc, CanonicalVectorView a, MasterVectorRef u) -> void
    {
        const auto dims = Mc.dims;
        const auto Rbs  = Mc.Rbs;
        const auto js   = Mc.js;
        const auto ju   = Mc.ju;

        const auto ns  = dims.ns;
        const auto nu  = dims.nu;
        const auto nbs = dims.nbs;
        const auto nl  = dims.nl;

        p.resize(dims.np);
        x.resize(dims.nx);
        auto xs = x.head(ns);
        auto xu = x.tail(nu);

        wbar.resize(dims.nw);
        auto wbs = wbar.head(nbs);

        const auto as = a.xs;
        const auto au = a.xu;
        const auto ap = a.p;
        const auto awbs = a.wbs.head(nbs);

        solveCanonical(Mc, {as, au, ap, awbs}, {xs, xu, p, wbs});

        w.noalias() = tr(Rbs) * wbs;

        u.x(js) = xs;
        u.x(ju) = au;
        u.p = p;
        u.w = w;
    }
};

LinearSolver::LinearSolver()
: pimpl(new Impl())
{}

LinearSolver::LinearSolver(const LinearSolver& other)
: pimpl(new Impl(*other.pimpl))
{}

LinearSolver::~LinearSolver()
{}

auto LinearSolver::operator=(LinearSolver other) -> LinearSolver&
{
    pimpl = std::move(other.pimpl);
    return *this;
}

auto LinearSolver::setOptions(const LinearSolverOptions& options) -> void
{
    pimpl->options = options;
}

auto LinearSolver::options() const -> const LinearSolverOptions&
{
    return pimpl->options;
}

auto LinearSolver::decompose(CanonicalMatrix Mc) -> void
{
    pimpl->decompose(Mc);
}

auto LinearSolver::solve(CanonicalMatrix Mc, MasterVectorView a, MasterVectorRef u) -> void
{
    pimpl->solve(Mc, a, u);
}

auto LinearSolver::solve(CanonicalMatrix Mc, CanonicalVectorView ac, MasterVectorRef u) -> void
{
    pimpl->solve(Mc, ac, u);
}

} // namespace Optima

File: tests/LinearSolver.py
# Optima is a C++ library for numerical solution of linear and nonlinear programing problems.
#
# Copyright © 2020-2024 Allan Leal
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.


from testing.optima import *
from testing.utils.matrices import *


tested_nx      = [15, 20, 30, 50]  # The tested number of x variables
tested_np      = [0, 5]            # The tested number of p variables
tested_ny      = [5, 8]            # The tested number of y variables
tested_nz      = [0, 5]            # The tested number of z variables
tested_nl      = [0, 2]            # The tested number of linearly dependent rows in Ax
tested_nu      = [0, 2]            # The tested number of unstable variables
tested_diagHxx = [False, True]     # The tested options for Hxx structure

# Tested cases for the linear solver methods
tested_methods = [
    LinearSolverMethod.Fullspace,
    LinearSolverMethod.Nullspace,
    LinearSolverMethod.Rangespace
]

@pytest.mark.parametrize("nx"     , tested_nx)
@pytest.mark.parametrize("np"     , tested_np)
@pytest.mark.parametrize("ny"     , tested_ny)
@pytest.mark.parametrize("nz"     , tested_nz)
@pytest.mark.parametrize("nl"     , tested_nl)
@pytest.mark.parametrize("nu"     , tested_nu)
@pytest.mark.parametrize("diagHxx", tested_diagHxx)
@pytest.mark.parametrize("method" , tested_methods)
def testLinearSolver(nx, np, ny, nz, nl, nu, diagHxx, method):

    params = MasterParams(nx, np, ny, nz, nl, nu, diagHxx)

    if params.invalid(): return

    dims = params.dims

    if method == LinearSolverMethod.Rangespace and not diagHxx:
        return  # Rangespace method only applicable to diagonal Hxx matrices - consider here only when both are true

    M = createMasterMatrix(params)

    nw = params.dims.nw

    uexp = MasterVector(dims)
    uexp.x = npy.linspace(1, nx, nx)
    uexp.p = npy.linspace(1, np, np)
    uexp.w = npy.linspace(1, nw, nw)

    a = M * uexp

    canonicalizer = Canonicalizer(M)

    Mc = canonicalizer.canonicalMatrix()

    options = LinearSolverOptions()
    options.method = method

    linearsolver = LinearSolver()
    linearsolver.setOptions(options)

    u = MasterVector(dims)

    linearsolver.decompose(Mc)
    linearsolver.solve(Mc, a, u)

    assert_almost_equal( (M * u).array(), a.array() )

    ju = M.ju  # the indices of the unstable variables in x

    assert all(u.x[ju] == a.x[ju])  # ensure ux[ju] == ax[ju]
